Mathlib/AlgebraicGeometry/EllipticCurve/NormalForms.lean

```diff
@@ -0,0 +1,658 @@
+  rw [EllipticCurve.j, Units.val_inv_eq_inv_val, ← div_eq_inv_mul, E.coe_Δ',
+    self.c₄_of_char_three, self.Δ_of_char_three, div_eq_div_iff h (right_ne_zero_of_mul h)]
+  ring1
+
+theorem j_ne_zero_of_char_three [CharP F 3] : E.j ≠ 0 := by
+  rw [self.j_of_char_three, div_ne_zero_iff]
+  have h := E.Δ'.ne_zero
+  rwa [E.coe_Δ', self.Δ_of_char_three, mul_ne_zero_iff] at h
+
+end IsCharThreeJNeZeroNF
+
+/-! ### Normal forms of characteristic = 3 -/
+
+/-- A `WeierstrassCurve` is in normal form of characteristic = 3, if it is
+$Y^2 = X^3 + a_2X^2 + a_6$ (`WeierstrassCurve.IsCharThreeJNeZeroNF`) or
+$Y^2 = X^3 + a_4X + a_6$ (`WeierstrassCurve.IsShortNF`). -/
+def IsCharThreeNF : Prop := W.IsCharThreeJNeZeroNF ∨ W.IsShortNF
+
+namespace IsCharThreeNF
+
+variable {W} (self : W.IsCharThreeNF)
+include self
+
+theorem isCharNeTwoNF : W.IsCharNeTwoNF :=
+  match self with
+  | Or.inl h | Or.inr h => h.isCharNeTwoNF
+
+end IsCharThreeNF
+
+section VariableChange
+
+variable [CharP R 3] [CharP F 3]
+
+/-- For a `WeierstrassCurve` defined over a ring of characteristic = 3,
+there is an explicit change of variables, which changes it to $Y^2 = X^3 + a_4X + a_6$
+(`WeierstrassCurve.IsShortNF`) if its j = 0.
+This is in fact given by `WeierstrassCurve.toCharNeTwoNF`. -/
+def toShortNFOfCharThree : VariableChange R :=
+  have h : (2 : R) * 2 = 1 := by linear_combination CharP.cast_eq_zero R 3
+  letI : Invertible (2 : R) := ⟨2, h, h⟩
+  W.toCharNeTwoNF
+
+/-- For a `WeierstrassCurve` defined over a field of characteristic = 3,
+there is an explicit change of variables of it to `WeierstrassCurve.IsCharThreeNF`, that is,
+$Y^2 = X^3 + a_2X^2 + a_6$ (`WeierstrassCurve.IsCharThreeJNeZeroNF`) or
+$Y^2 = X^3 + a_4X + a_6$ (`WeierstrassCurve.IsShortNF`).
+It is the composition of an explicit change of variables with
+`WeierstrassCurve.toShortNFOfCharThree`. -/
+def toCharThreeNF (W : WeierstrassCurve F) : VariableChange F :=
+  ⟨1, (W.variableChange W.toShortNFOfCharThree).a₄ /
+    (W.variableChange W.toShortNFOfCharThree).a₂, 0, 0⟩ * W.toShortNFOfCharThree
+
+lemma toShortNFOfCharThree_a₂ : (W.variableChange W.toShortNFOfCharThree).a₂ = W.b₂ := by
+  simp_rw [toShortNFOfCharThree, toCharNeTwoNF, variableChange_a₂, inv_one, Units.val_one, b₂]
+  linear_combination (-W.a₂ - W.a₁ ^ 2) * CharP.cast_eq_zero R 3
+
+theorem toShortNFOfCharThree_spec (hb₂ : W.b₂ = 0) :
+    (W.variableChange W.toShortNFOfCharThree).IsShortNF := by
+  have h : (2 : R) * 2 = 1 := by linear_combination CharP.cast_eq_zero R 3
+  letI : Invertible (2 : R) := ⟨2, h, h⟩
+  have H := W.toCharNeTwoNF_spec
+  exact ⟨H.a₁, hb₂ ▸ W.toShortNFOfCharThree_a₂, H.a₃⟩
+
+section
+
+variable (W : WeierstrassCurve F)
+
+theorem toCharThreeNF_spec_of_b₂_ne_zero (hb₂ : W.b₂ ≠ 0) :
+    (W.variableChange W.toCharThreeNF).IsCharThreeJNeZeroNF := by
+  have h : (2 : F) * 2 = 1 := by linear_combination CharP.cast_eq_zero F 3
+  letI : Invertible (2 : F) := ⟨2, h, h⟩
+  rw [toCharThreeNF]; erw [variableChange_comp]
```

Comment:
> Do you think this is just due to missing API?

Yes. I have a plan to refactor `VariableChange` to in favor of `*` instead of `VariableChange.comp`.